httpClient

chris01b opened the pull request:
# HTTP Client Refactor & Auth Enhancements

## Summary

This pull request introduces a significant refactoring of the frontend's HTTP communication layer by implementing a new, `HttpClient` based on Axios. It also includes substantial enhancements to backend authentication mechanisms, focusing on CSRF protection, token refresh, security headers, and standardized error handling. Comprehensive unit tests for the new frontend HTTP client and updates to backend tests are included.

A new secret for staging/production is required `JWT_SECRET_KEY`.

## Changes

### Frontend

1.  **New `HttpClient` Service**:
    *   Replaces the previous `fetch`-based `makeRequest` utility.
    *   Built on Axios with features like:
        *   Centralized configuration (baseURL, timeout, `withCredentials`).
        *   Automatic request retries using `axios-retry`.
        *   **Request Interceptor**:
            *   Injects `Authorization: Bearer <id_token>` from cookies.
            *   Injects `X-XSRF-TOKEN` header from cookies for state-changing requests, ensuring CSRF protection.
        *   **Response Interceptor**:
            *   Handles 401/422 errors by attempting to refresh the ID token via the new `/api/auth/refresh-token` backend endpoint.
            *   Queues concurrent requests during token refresh and retries them with the new token.
        *   **Error Normalization**: Converts Axios, network, and other errors into a consistent `HttpError` object, providing structured details (`apiError` property) like status, backend error code, and message.
    *   Located in `frontend/src/lib/http.ts`, with helpers in `http.helpers.ts` and types in `http.types.ts`.

2.  **`HttpClientContext` and `useHttpClient` Hook**:
    *   A React Context (`frontend/src/lib/HttpClientContext.tsx`) provides the `HttpClient` instance throughout the application.
    *   The `useHttpClient` hook offers a convenient API (`request`, `requestRawResponse`, and shortcuts like `get`, `post`) for components and other hooks.
    *   The `Root` component in `frontend/src/router.tsx` now wraps the application with `HttpClientProvider`.

3.  **Codebase-Wide Refactoring**:
    *   All frontend components, contexts, and utility functions previously using `makeRequest` have been updated to use the new `useHttpClient` hook or the direct `httpClient` instance.
    *   Error handling logic in components is updated to utilize the richer information from `HttpError` for displaying user-friendly messages.

4.  **Enhanced Testing**:
    *   Added `axios-retry`, `@testing-library/react`, `vitest`, and `@vitest/coverage-v8` as dev dependencies.
    *   Configured ESLint, Vite (`jsdom` environment, globals), and TypeScript (`vitest/globals` types, `tests` directory include) for Vitest.
    *   Comprehensive unit tests for:
        *   `HttpClientContext` and `useHttpClient` hook (`frontend/tests/lib/HttpClientContext.test.tsx`).
        *   HTTP helper functions (`frontend/tests/lib/http.helpers.test.ts`).
        *   The `HttpClient` class, including interceptors and token refresh logic (`frontend/tests/lib/http.test.ts`).

### Backend

1.  **Enhanced Security Headers**:
    *   `backend/app.py` now includes an `after_request` hook to add `Content-Security-Policy`, `X-Content-Type-Options`, `X-Frame-Options`, and `Strict-Transport-Security` headers to all responses.

2.  **Improved Authentication and Authorization**:
    *   **Standardized Error Responses**:
        *   `create_error_response` (in `backend/auth/utils/responses.py`) now primarily uses a `message_key` argument.
        *   This key is used to look up standardized error messages and corresponding error codes from the greatly expanded and categorized `AUTH_ERROR_MESSAGES` dictionary in `backend/auth/providers/cognito/constants.py`.
        *   Auth controllers and providers are updated to use this new system.
    *   **CSRF Protection via Flask-JWT-Extended**:
        *   The deprecated custom `auth_required` decorator is removed.
        *   `participant_auth_required` and `researcher_auth_required` decorators now integrate `verify_jwt_in_request()` from `flask_jwt_extended` for CSRF protection.
        *   `backend/config.py` includes new and updated JWT configurations (secret key, token location in cookies, CSRF protection flag, cookie/header names like `XSRF-TOKEN`, CSRF methods, and environment-aware `Secure` and `SameSite` attributes for cookies).
        *   CORS configuration (`CORS_ALLOW_HEADERS`, `CORS_EXPOSE_HEADERS`) updated to include `X-XSRF-TOKEN`.
    *   **Token Refresh Endpoint**:
        *   A new endpoint `/api/auth/refresh-token` is added in `backend/views/api/auth.py`.
        *   It uses the `refresh_token` (from cookies) to obtain new `id_token` and `access_token` from Cognito.
        *   User type (participant/researcher) is determined from the existing ID token's issuer.
        *   `CognitoAuthBase` in `backend/auth/providers/cognito/base.py` includes a new `refresh_id_token` method to facilitate this.
    *   **Cookie Management (`backend/auth/utils/cookies.py`)**:
        *   `clear_auth_cookies` now also clears the `XSRF-TOKEN` cookie.
        *   `set_auth_cookies` is significantly enhanced to:
            *   Set `id_token`, `access_token`, and `refresh_token` cookies with appropriate `HttpOnly`, `Secure` (environment-aware), and `SameSite` (environment-aware) attributes.
            *   Utilize `create_access_token` and `set_access_cookies` from Flask-JWT-Extended to manage the double-submit CSRF cookies (`XSRF-TOKEN` and its associated HttpOnly JWT).
            *   Expose the new CSRF token in the `X-XSRF-TOKEN` response header, allowing SPAs to update their cached token.

3.  **Backend Test Enhancements**:
    *   `tests/conftest.py` introduces a new session-scoped autouse fixture `_patch_jwt_functions`. This fixture mocks various `flask_jwt_extended` functions and auth-related functions within the application to bypass JWT/CSRF checks during unit tests, simplifying the testing of application logic.
    *   Unit tests for auth providers (`test_participant_auth.py`, `test_researcher_auth.py`) and cookie utilities (`test_cookies.py`) are updated to align with new error messages and cookie behaviors.

## How To Test

**General Frontend Functionality:**

1.  **Log in as a Researcher and Participant**: Verify successful login and basic navigation for both roles.
2.  **Data Display**: Ensure all data loads correctly in tables, forms, and summaries across all dashboards.
3.  **CRUD Operations**:
    *   **Admin Dashboard**: Test create, edit, and archive/delete for About Sleep Templates, Access Groups, Accounts, Roles, Studies. Test Data Retrieval Task invocation and stopping.
    *   **Ditti App**: Test audio file uploads, deletion, and subject detail editing.
    *   **Participant Dashboard**: Test study consent flow.
    *   Confirm operations succeed and UI updates correctly. Verify appropriate success/error flash messages.

**Specific HTTP Client & Auth Scenarios:**

1.  **CSRF Protection**:
    *   Using browser developer tools, attempt to tamper with or remove `X-XSRF-TOKEN` from state-changing requests. Operations should fail (likely 401/403).
    *   Confirm `XSRF-TOKEN` cookie and `X-XSRF-TOKEN` request header are present on relevant (e.g., POST, PUT) requests.

2.  **Token Refresh**:
    *   Login then edit JWT_SECRET_KEY in `config.py` while Flask is running:
        *   The next API call should initially fail (401/422).
        *   A call to `/api/auth/refresh-token` should be made automatically.
        *   New `id_token`, `access_token`, and `XSRF-TOKEN` cookies should be set.
        *   The `X-XSRF-TOKEN` header should be present in the refresh response.
        *   The original request should be retried successfully with the new tokens.
    *   If the refresh token itself is expired, refresh should fail, and the user should effectively be logged out (persistent 401s).

3.  **Error Handling**:
    *   **Network Disconnection**: Simulate network loss and attempt an API call. Verify a user-friendly network error message.
    *   **Backend Errors**: Trigger known backend validation errors or force server errors. Verify the frontend displays the specific error message and code from the backend.
    *   Check browser console for detailed `HttpError` objects.

4.  **Request Retries**:
    *   (Difficult to test reliably) If transient network issues or intermittent 5xx errors for idempotent requests can be simulated, observe if `HttpClient` retries.

**Backend Testing**:

1.  **Security Headers**: Inspect HTTP response headers in the browser to confirm the presence of `Content-Security-Policy`, `X-Content-Type-Options`, etc.
2.  **Unit Tests**: Ensure all backend unit tests pass.

## Checklist
- [x] New code includes a license statement.
- [x] Any TODOs are made into new issues.
- [x] Code is properly linted (for Markdown, TypeScript, and Python).
- [x] Unit tests are written for any new code.
- [x] All existing unit tests pass.
- [x] Any changes to the database schema include a migration script that does not result in data loss (if applicable).
- [x] Any new endpoints include proper AWS Cognito and/or RBAC protections (if applicable).

chris01b committed:
Update axios, add axios-retry

chris01b committed:
Add HttpClient

chris01b committed:
Add optional code key to ResponseBody

chris01b committed:
utils.tsx -> utils.ts, remove makeRequest, use httpClient.request instead

chris01b committed:
Remove unused custom RequestConfig

chris01b committed:
Update implementations

chris01b committed:
Added requestRawResponse method to HttpClient, added comment about X-CSRF-TOKEN being unused

chris01b committed:
Refactor downloadExcelFromUrl to use httpClient.requestRawResponse() with better error handling

chris01b committed:
Remove unnecessary comment

chris01b committed:
Fix bug introduced by changing /data_processing_task/force-stop's parameters, revert to original handleForceStop loading conditions

JStover95 reviewed (changes requested):
This is much needed. I only left two comments.

It would also be a good idea for us to start including tests on our frontend code. I have used `mocha` in the past and am open to other recommendations.

chris01b committed:
Remove localStorage JWT handling, comment line length <= 80

Authentication is now handled by same-site strict cookies, so JWT storage and retrieval via localStorage is no longer necessary.

This commit removes:
- Retrieving JWT from localStorage in request interceptor.
- Setting Authorization header with JWT from localStorage.
- Storing JWT from response body into localStorage.

chris01b committed:
Refactors the frontend HTTP client usage by introducing a Context provider and a `useHttpClient` hook.

Key changes:
- Created `HttpClientContext.tsx` defining `HttpClientProvider` and `useHttpClient`.
- Modified `http.ts` to export both the `HttpClient` class (for the provider) and the `httpClient` singleton instance (for non-React utility files).
- Updated `index.tsx` to wrap the application with `HttpClientProvider`.
- Refactored numerous React components, contexts, and hooks to use the `useHttpClient` hook instead of importing the `httpClient` singleton directly.
- Utility files (`utils.ts`, `dataFactory.ts`) continue to use the `httpClient` singleton to avoid excessive prop drilling or complex refactoring.

This change decouples React components from the specific HTTP client implementation, enabling easier mocking in tests and providing a more flexible foundation for future enhancements like centralized error handling or request cancellation.

chris01b committed:
Add the Vitest Explorer extension to the recommended VSCode extensions

chris01b committed:
Add TypeScript/TSX specific ESLint rule overrides for test files to allow common testing patterns

chris01b committed:
Add Vitest, Testing Library, and related dependencies; add test and coverage scripts to package.json; include Vitest globals and the tests directory in tsconfig.json; configure Vitest in vite.config.js

chris01b committed:
Add unit tests for the HttpClient utility and the HttpClientProvider/useHttpClient context and hook

chris01b committed:
Merge branch 'dev' into httpClient

chris01b committed:
Add license to new tests

chris01b committed:
Uninstall unplugin-auto-import

chris01b committed:
Add universal security headers

chris01b committed:
Unify CORS + JWT/CSRF configuration

chris01b committed:
Centralize AUTH_ERROR_MESSAGES helpers

chris01b committed:
New `create_error_response` signature

chris01b committed:
Hardened cookie helpers & CSRF double-submit JWT

chris01b committed:
Added `/api/auth/refresh-token` endpoint

chris01b committed:
Drop deprecated JWT decorator and add CSRF verify

chris01b committed:
Consistent error handling & redirect helpers

chris01b committed:
Database/validation guardrails

chris01b committed:
Token-aware Axios client with silent refresh queue

chris01b committed:
Provide HttpClient via context provider

chris01b committed:
Standardize CSRF/XSRF naming and clean up frontend logic

- Switched all CSRF cookie and header names to Axios defaults:
  - Cookie: XSRF-TOKEN
  - Header: X-XSRF-TOKEN
- Updated Flask-JWT-Extended config and all CORS allow/expose headers to match new names.
- Updated backend CSRF cookie clearing and header exposure to use new names.
- Refactored frontend http.ts:
  - Removed all legacy/manual CSRF token handling and references to Flask-WTF.
  - Now always sets X-XSRF-TOKEN header for state-changing requests using the XSRF-TOKEN cookie.
  - Token refresh now uses JSON and relies on Axios to attach the correct CSRF header.
  - Cleaned up and clarified comments.

chris01b committed:
Ruff format

chris01b committed:
Refactor HTTP client and enhance React context API

- Add HTTP verb shortcuts (get, post, put, delete, patch) to HttpClientContext API
- Extract helper functions from HttpClient class for better code organization
- Remove redundant method implementations from HttpClient class
- Improve documentation with JSDoc comments and usage examples

chris01b committed:
Continued

chris01b committed:
Improve HttpClient test suite organization and type safety

- Add helper functions to reduce code duplication in test setup
- Improve type safety by adding proper interfaces for test mocks
- Update CSRF token handling tests to use cookies instead of localStorage
- Add comprehensive tests for token refresh and queue processing
- Fix error handling tests to properly check normalized HttpError instances
- Add tests for login redirect behavior based on URL path
- Clean up test setup and teardown with proper beforeEach/afterEach hooks

chris01b committed:
Add http.helpers.test.ts

chris01b committed:
Add HTTP verb method tests to HttpClientContext

- Add tests for all HTTP verb methods (GET, POST, PUT, DELETE, PATCH)
- Update TestConsumerComponent to support all HTTP methods
- Add test cases for method-specific configurations and data handling
- Verify correct request parameters and method calls
- Test response handling for each HTTP verb

chris01b committed:
Update pytests

chris01b committed:
Remove unused dependency @testing-library/jest-dom

chris01b committed:
Refactor HttpClient to hooks and contexts folders, update tests, update uses

chris01b committed:
Update old comment

JStover95 commented on backend/app.py:
```diff
@@ -74,6 +75,26 @@ def log_response(response: Response):
+            "default-src 'self'; img-src 'self' data:; "
+            "script-src 'self' 'unsafe-inline'; "
+            "style-src 'self' 'unsafe-inline'; font-src 'self';"
+        )
```
I did some preliminary research into a good CSP header for our stack, and this may need some adjustments. Let's discuss in more details when we meet next.

JStover95 commented on backend/auth/utils/responses.py:
```diff
@@ -12,8 +12,15 @@
+
+def create_error_response(
+    message=None, status_code=401, error_code=None, message_key=None
+):
```
Since we now have many potential messages, I recommend defining a `ErrorResponseMessageKey` string literal in a separate `backend/auth/providers/cognito/types.py` files and typing this `create_error_response` function. For example:

```py
type ErrorResponseMessageKey = Literal["auth_failed", "auth_required", ...]
```

chris01b replied:
Added more types to this file

JStover95 commented on frontend/src/lib/http.helpers.ts:
Please add a license statement to this file.

chris01b replied:
Added

JStover95 commented on frontend/src/utils.ts:
```diff
@@ -289,39 +234,37 @@ export const makeRequest = async (
 export async function downloadExcelFromUrl(
```
We should replace `fetch` in this function with the `httpClient`. This can be done in a future PR.

JStover95 reviewed (changes requested):
Looks great! Just a few comments. Let's discuss today.

JStover95 commented:
I also had to add a few changes to get vitest to work as expected. Please add these changes as well.

1. Add `vitest.config.js`:

   ```js
   import { defineConfig } from "vitest/config";
   
   export default defineConfig({
     test: {
       projects: ["src/*"],
       globals: true,
       environment: "jsdom",
     },
   });
   ```

2. Add `"@typescript-eslint/no-unsafe-return": "off"` to `rules` in `eslint.config.js`.
3. Add `"vitest.config.js"` to `include` in `tsconfig.json`.

JStover95 commented on backend/auth/utils/cookies.py:
```diff
@@ -81,9 +106,22 @@ def set_auth_cookies(response, token):
+
+    # This helper sets BOTH the HttpOnly JWT cookie and the non-HttpOnly
+    # CSRF cookie (named according to our JWT_* config options).
+    set_access_cookies(response, csrf_jwt)
```
I noticed this is setting an unused `access_token_cookie`. It would be better for us to set the csrf token directly instead of using this function:

```py
from flask_jwt_extended.config import config

response.set_cookie(
    config.access_csrf_cookie_name,
    value=get_csrf_token(csrf_jwt),
    max_age=config.cookie_max_age,
    secure=config.cookie_secure,
    httponly=False,
    domain=config.cookie_domain,
    path=config.access_csrf_cookie_path,
    samesite=config.cookie_samesite,
)
```

chris01b committed:
Add better types to responses.py

chris01b committed:
Add license to http.helpers.ts

chris01b committed:
Add vitest.config.js and add to tsconfig.json

chris01b committed:
Revert "Add vitest.config.js and add to tsconfig.json"

This reverts commit f931dfbd128a04b7651f0311d431cd49af5c1089.

chris01b committed:
Set Dockerfile zappa version to 0.59.0

chris01b committed:
Enhance Content Security Policy and security headers

- Development/Testing: More permissive CSP allowing inline scripts/styles, eval, and Vite dev server connections for ease of development.
- Staging/Production: Stricter CSP, restricting sources to 'self' and the configured CloudFront domain. Includes directives for frame-ancestors, form-action, base-uri, and object-src for enhanced security.

chris01b committed:
Set JWT_COOKIE_DOMAIN for staging and production

This allows cookies set by the backend to be accessible to the frontend on a different subdomain, which is crucial for cross-subdomain CSRF token handling and general cookie-based authentication.

chris01b committed:
Ignore local staging frontend environment variables

chris01b committed:
Configure Vite build for production and staging

- Enables manifest generation by default.
- Configures `rollupOptions.output` for consistent asset naming (entry, chunk, asset files) with hashes for cache busting.
- Sets `assetsInlineLimit` to 0 to ensure assets are not inlined.
- Enables `cssCodeSplit`.
- Disables `sourcemap` for production/staging builds by default.

chris01b committed:
Ensure consistent cookie attributes for auth and CSRF tokens

- `set_auth_cookies`: Now uses `JWT_COOKIE_DOMAIN` and `JWT_COOKIE_PATH` from app config when setting `id_token`, `access_token`, and `refresh_token` cookies. This aligns them with Flask-JWT-Extended's CSRF cookie attributes.
- `clear_auth_cookies`:
- Uses `flask-jwt-extended`'s `unset_access_cookies` and `unset_refresh_cookies` to correctly clear JWT-managed cookies (including the primary CSRF token), respecting all app config for domain, path, secure, and samesite.
- Explicitly clears custom cookies (`id_token`, `access_token`, `refresh_token`) using attributes derived from app config for consistency.

chris01b committed:
Specify staging mode for frontend build script

This ensures that Vite loads the correct environment configuration (e.g., from `.env.staging.local`) when building for the staging environment.

chris01b committed:
Enhance cookie attribute configuration and JWT testing

- The `set_auth_cookies` utility in `backend/auth/utils/cookies.py` now prioritizes `JWT_COOKIE_SECURE` and `JWT_COOKIE_SAMESITE` from the Flask app configuration when setting the `Secure` and `SameSite` attributes on authentication cookies. It falls back to inferring these based on the `current_app.debug` or `current_app.testing` status if these configuration values are not explicitly set.
- Enhanced JWT Cookie Testing:
  - The tests in `tests/tests_auth/utils/test_cookies.py` have been significantly refactored for greater accuracy and coverage.
  - The `mock_app` fixture now initializes `JWTManager` with a comprehensive set of JWT-related configurations (e.g., `JWT_COOKIE_SECURE`, `JWT_COOKIE_SAMESITE`, cookie names, paths), providing a more realistic testing environment for cookie behavior.
  - Tests for `set_auth_cookies` and `clear_auth_cookies` now perform more precise assertions. They verify not only the presence of cookies but also their attributes (`HttpOnly`, `Secure`, `SameSite`, `Path`, `Domain`) strictly against the application's JWT settings.
  - Assertions now cover standard Flask-JWT-Extended cookies (access token cookie, CSRF token cookie) in addition to custom application cookies (`id_token`, `access_token`, `refresh_token`).
- Several broad mocks for `flask_jwt_extended` core functions and related local utility functions have been removed from `tests/conftest.py`. This change aims to make tests rely more on the actual JWT extension's behavior or more targeted, specific mocks where necessary, improving the overall reliability of authentication tests.

chris01b committed:
Merge branch 'dev' into httpClient

JStover95 reviewed (approved):
Looks good!
